backends/metad/client.go

```diff
@@ -98,31 +84,48 @@ func NewMetadClient(backendNodes []string) (*Client, error) {
 }
 
 func (c *Client) selectConnection() error {
-	//random start
-	if c.current == nil {
-		rand.Seed(time.Now().Unix())
-		r := rand.Intn(c.connections.Len())
-		c.connections = c.connections.Move(r)
-	}
-	c.connections = c.connections.Next()
-	conn := c.connections.Value.(*Connection)
-	startConn := conn
-	err := conn.testConnection()
-	for err != nil {
-		log.Error("Connection to [%s], error: [%s]", conn.url, err.Error())
+	var err error
+	maxTime := 15 * time.Second
+	i := 1 * time.Second
+	for ; i < maxTime; i *= time.Duration(2) {
+
+		//random start
+		if c.current == nil {
+			rand.Seed(time.Now().Unix())
+			r := rand.Intn(c.connections.Len())
+			c.connections = c.connections.Move(r)
+		}
 		c.connections = c.connections.Next()
-		conn = c.connections.Value.(*Connection)
-		if conn == startConn {
-			return errors.New("Fail to connect any backend.")
+		conn := c.connections.Value.(*Connection)
+		startConn := conn
+
+		//iterate through connection
+		_, err = conn.makeMetaDataRequest("/")
+		for err != nil {
+			log.Error("Connection to [%s], error: [%s]", conn.url, err.Error())
+			c.connections = c.connections.Next()
+			conn = c.connections.Value.(*Connection)
+			if conn == startConn {
+				break
+			}
+			_, err = conn.makeMetaDataRequest("/")
```

Comment:
logic is complex

Reply:
move the inner loop to testConnection function.